### ISSUE TITLE
Multikey Bindings

### ISSUE THREAD

**Issue by noctuid** [arybczak/ncmpcpp#24]
I thought I saw this as an issue before but apparently not. Being able to bind key sequences like "gg" or "space key" to actions is pretty essential for me. Is this something you'd be willing to consider adding?

**Comment by arybczak** [comment]
The problem with multikey bindings is that you need to have a delay during which you wait for for the possible next key in the sequence. Since you don't want this delay to fire at all times, you'd probably need additional key that enables this delay and this resembles commands system that is already in place (you can press ':' to enter the mode of typing a command, set commands "gg", " k" etc. and make them execute as soon as they're typed, without the need to type enter afterwards.

**Comment by angelic-sedition** [comment]
Are you talking about a timeout where you could have both "gg" and "g" bound, and if you wanted to use "g", you would have to wait for the timeout? That's just one way to implement them. Vim, for example, also lets you get rid of the timeout. If there's no possible next key, there's no need for a delay. Unlike with a command system, you could not have a key sequence that was a subset of another (e.g. you could not have both "cat" and "caterpillar"). Every sequence must be unique. If you have "gg" bound, "g" could not also be bound to something. If "ggg" is bound, "gg" can not also be bound. Not having a timeout actually seems to be more common among what I use; most people I've talked with seem to just find the delay annoying and not worth 1 extra keybind.

Basically after a keypress that has a mapping at a further depth, there is an indenfinite wait for the next keypress until the user reaches the max depth or cancels it with ESC, in which case ESC cannot be bound in a sequence (at least not without more complex checks). Once the last depth is reached, if the key is bound to an action, it is taken. Otherwise, the sequence is just cancelled (e.g. if you have "gg" bound and press "ga", nothing will happen). Vim actually will execute the previous level's action even with 0 timeout if you hit a key not bound (e.g. ESC), but I find this annoying and overly complicated as well. I think it would be preferable to either ignore a "g" mapping entirely if "gg" is bound or to have the user explicitly unmap "g" first, which is how mpv recently implemented this.

What do think about this method?

**Comment by arybczak** [comment]
> If there's no possible next key, there's no need for a delay.

Sure, that's a way to refine it.

> Basically after a keypress that has a mapping at a further depth, there is an indenfinite wait for the next keypress until the user reaches the max depth or cancels it with ESC

So then if you bind gg and ggg, you essentially have no way to invoke gg, because after gg it will wait indefinitely for the next g?

> I think it would be preferable to either ignore a "g" mapping entirely if "gg" is bound or to have the user explicitly unmap "g" first, which is how mpv recently implemented this.

> I think it would be preferable to either ignore a "g" mapping entirely if "gg" is bound or to have the user explicitly unmap "g" first, which is how mpv recently implemented this.

So now (as opposed to using commands for this) instead of accepting only the shortest sequence you accept only the longest one. I can't see how this is better.

The bottom line is that the way ncmpcpp collects input from the keyboard is currently stateless. I recently was loooking for a way to allow meta_X key combinations and found out that it needs support for intermediate state, just as multi key bindings. I'm personally not interested enough in this problem to solve it myself, but anyone who wishes to tackle it is welcome to try.

**Comment by angelic-sedition** [comment]
> So then if you bind gg and ggg, you essentially have no way to invoke gg, because after gg it will wait indefinitely for the next g?

If you bound ggg, you wouldn't want to invoke gg. I think it unlikely that anyone would need so many mappings that two key mappings using only a few prefix keys wouldn't suffice. 

> So now (as opposed to using commands for this) instead of accepting only the shortest sequence you accept only the longest one. I can't see how this is better.

It saves keypresses and allows for a lot more easily pressable mappings. Commands are different. :gg\<cr\> is 4 key presses. gg is 2. You can no longer map just g, but you can map ga, gb, etc., more than making up for a single lost binding.

Thanks for considering it. I agree it may not be very enticing given the amount of work needed to implement them.

**Comment by arybczak** [comment]
> It saves keypresses and allows for a lot more easily pressable mappings. Commands are different. :gg<cr> is 4 key presses. gg is 2. You can no longer map just g, but you can map ga, gb, etc., more than making up for a single lost binding.

The thing is that with commands you need to press :gg instead of gg (pressing enter in the end is not necessary if you define the command as immediate), which is possibly still inconvenient, but less so.

**Comment by angelic-sedition** [comment]
In that case, I didn't realize the command system allowed for 'immediate'. When using immediate instead of deferred, this is the same functionality I'm suggesting. For example, one could bind g to the command key and then make g an immediate command.

The difference is that this would only allow for one prefix key. Someone could also bind \<space\> to `execute_command`, but it would just behave the same since you can only have one set of commands. Do you think you could extend the current command system to allow for more than just one key that can have this behaviour? The functionality is the same. I could, for example, probably just duplicate code to make a `def_command2`, `run_command2`, `def_command3`, etc. to allow for for multiple prefix keys where the "commands"/keys took different actions, but that would be pretty ugly. I don't think it's very important that only the shortest sequence would be accepted; that's up to the user.

**Comment by quite** [comment]
@mjessome thats really cool

**Comment by arybczak** [comment]
@angelic-sedition Can you try the patch referenced above and tell me about its usability, i.e. if it's clear how to define multikey binding, how well does it work etc.?

**Comment by angelic-sedition** [comment]
Multikey binding show up in the help fine (F1). Defining multikey bindings is pretty much the same:

```
def_key "gg"
 move_home
```
The description  sounds nice, but it does not work properly for me. With the above keybinding, pressing "gg" within the timeout works. Taking the example in #29, if I have both g ang gg bound with the default timeout and press "gz" within the timeout, nothing happens as expected. However, if I press "g" and then wait for the timeout, the "g" action is not performed after the timeout. The "g" action will happen, but only after the next keypress. It is implied that more than 2 key sequences should work ("prefix of gz..."), but I also couldn't get sequences of 3 keys working (e.g. def_key "ggg"). Sometimes it seems that things get stuck so that no matter what keys I press, nothing happens.

**Comment by mjessome** [comment]
@angelic-sedition 
1. I can reproduce the issue of keystrokes not registering on timeout. I'll look into it.
2. I can't reproduce the 3 key issue: I bound 'g', 'gg', 'gz', 'ggg', 'ggz', 'gzz' and all 3-key bindings work without issue. The 1 and 2 key bindings don't register on timeout as per (1) above.
3. I can't reproduce things "getting stuck".

I'll look into all of this tomorrow, thanks for giving it a try. I think the timeout issue is due to the way the main loop waits for keystrokes, and it should be easy enough to fix.

**Comment by arybczak** [comment]
@mjessome  any progress?

**Comment by tremby** [comment]
Was it implemented, or...?

**Comment by xeruf** [comment]
Yeah, the PR was just closed and I do not see why this was closed too?

**Comment by ifohancroft** [comment]
I'd love this feature as well.

Currently, on ncmpcpp version 0.9.2 on Manjaro installed from the repo, trying to bind 'gg' gives an error on ncmpstart.

The binding:

```
def_key "gg"
  move_home
```

The error:

```
$ ncmpcpp 
Reading configuration from /home/ifohancroft/.config/ncmpcpp/config...
/home/ifohancroft/.config/ncmpcpp/bindings:192: error: invalid key: 'gg'
```

**Comment by gavieira** [comment]
Yep, still having the same issue as @ifohancroft
